File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int preOrderIndex = 0;
    public int findPosition(int inorder[], int element, int n){
        for(int i = 0; i < n; i++)
            if(inorder[i] == element)
                return i;
        return -1;
    }
    public TreeNode solve(int inorder[], int preorder[], int n, int inOrderStart, int inOrderEnd){
        if(preOrderIndex >= n || inOrderStart > inOrderEnd)
            return null;
        
        int element = preorder[preOrderIndex++];
        TreeNode root = new TreeNode(element);
        int position = findPosition(inorder, element, n);
        
        root.left = solve(inorder, preorder, n, inOrderStart, position-1);
        root.right = solve(inorder, preorder, n, position+1, inOrderEnd);
        
        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = preorder.length;
        TreeNode ans = solve(inorder, preorder, n, 0, n-1);
        return ans;
    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* solve(vector<int>& preorder, int preLow, int preHigh, vector<int>& inorder, int inLow, int inHigh){
        if(inLow > inHigh && preLow > preHigh)
            return NULL;
        TreeNode* root = new TreeNode(preorder[preLow]);

        int index = inLow;
        while(index <= inHigh){
            if(inorder[index] == preorder[preLow])
                break;
            index++;
        }

        root->left = solve(preorder, preLow+1, preLow+index-inLow, inorder, inLow, index-1);
        root->right = solve(preorder, preLow+index-inLow+1, preHigh, inorder, index+1, inHigh);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return solve(preorder, 0, preorder.size()-1, inorder, 0, inorder.size()-1);
    }
};